libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include "AP_Beacon_Nooploop.h"
+#include <ctype.h>
+#include <stdio.h>
+
+#define NOOPLOOP_INVALID_VAL -8388000
+
+extern const AP_HAL::HAL& hal;
+
+AP_Beacon_Nooploop::AP_Beacon_Nooploop(AP_Beacon &frontend, AP_SerialManager &serial_manager) :
+    AP_Beacon_Backend(frontend)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Beacon, 0));
+    }
+}
+
+// return true if sensor is basically healthy (we are receiving data)
+bool AP_Beacon_Nooploop::healthy()
+{
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - _last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+// update the state of the sensor
+void AP_Beacon_Nooploop::update(void)
+{
+    // return immediately if not serial port
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check uart for any incoming messages
+    uint32_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        uint8_t b = _uart->read();
```

Comment:
Pretty sure this returns an `int16_t` - notionally it can fail (current implementations won't with this setup).  But should cope with it here.

Reply:
I think this issue is also still outstanding?

Reply:
Thank you very much. But may I ask is it really necessary? I found every driver simply cast it to uint8_t and use it.

Reply:
```
libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp:64:        const int16_t b = uart->read();
libraries/AP_RangeFinder/AP_RangeFinder_Benewake.cpp:62:        int16_t r = uart->read();
libraries/AP_RangeFinder/AP_RangeFinder_Lanbao.cpp:48:        int16_t b = uart->read();
libraries/AP_RangeFinder/AP_RangeFinder_LeddarVu8.cpp:54:        int16_t r = uart->read();
libraries/AP_Winch/AP_Winch_Daiwa.cpp:106:        int16_t b = uart->read();
libraries/RC_Channel/examples/RC_UART/RC_UART.cpp:39:        int16_t c = hal.UART->read();
```

Those were the ones at the bottom of a simple grep.

Reply:
My fault. Thank you very much. It is fixed.

Reply:
I agree there are some drivers that don't follow this practice but they probably should.  We are slowly "upping our game".. i.e. improving the safety and quality of our drivers.